### ISSUE TITLE
[Question] Working with autocomplete and foreign languages

### ISSUE THREAD

**Issue by mfxuus** [yourlabs/django-cities-light#180]
I've implemented the cities-light + autocomplete in English and everything is working well. Now I'm trying to do 2 things:

1. display names in foreign languages. I've read somewhere that I could use django-modeltranslation to help that out, so I will explore that further;

2. enable my users to search/autocomplete when they type in foreign characters (in particular, simplified Chinese). I've been digging around, experimenting with different approaches. For example, I edit/modify fields 'name', 'display_name', 'search_name', etc , and it seems that (for City) the autocomplete is searching against the 'name' column. In particular, when I change the 'name' field for Shanghai to '上海' (Chinese for Shanghai), and I type in "上" on the web form, the autocomplete returns "Shanghai, Shanghai, China" (the display name) as expected. 

However, I want the English searching/autocompleting to keep working as before. So, I tried changing the 'name' column to '上海Shanghai' / '上海 Shanghai' / 
'上海 shanghai' ,etc. , but regardless of the how I put the second part, the English autocomplete does not work (if I type "shang", Shanghai does not get returned). 

So I continued experimenting. If I make the 'name' field something like  'Shanghai上海' (English first, Chinese afterwards), the English autocomplete works, but the Chinese one stops working. It feels almost like it's searching with a 'begins_with" filter ... I'm sure it's not, but I'm not sure what's happening...

In summary, I would appreciate very much any advice on how to proceed with the two challenges I face: 1. displaying cities/countries in different languages (is modeltranslation the way to go?); and 2. Kindly walk me through what autocomplete is searching against here, and what I should do to enable the autocomplete for cities in both Chinese and English.

Thanks!

**Comment by jpic** [comment]
Can you extract SQL query log and copy it here ?

**Comment by mfxuus** [comment]
@jpic Hmm do you mean the log where I experiment with the different ''name'' values? I'm using DB Browser for SQLite and there are a couple of lines of these I see:

```
UPDATE `cities_light_city` SET `name`=? WHERE _rowid_='23093';
UPDATE `cities_light_city` SET `name`=? WHERE _rowid_='23093';
```

and this is what shows on my server log:

```
[29/Apr/2018 17:33:42] "GET /city-autocomplete/?q=shang&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 418
[29/Apr/2018 17:33:49] "GET /city-autocomplete/?q=shanghai&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 98
[29/Apr/2018 17:33:51] "GET /city-autocomplete/?q=shang&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 418
[29/Apr/2018 17:33:53] "GET /city-autocomplete/?q=s&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 610
[29/Apr/2018 17:33:54] "GET /city-autocomplete/?q=&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 663
[29/Apr/2018 17:33:55] "GET /city-autocomplete/?q=%E4%B8%8A&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 46
[29/Apr/2018 17:34:05] "GET /city-autocomplete/?q=&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 663
[29/Apr/2018 17:34:06] "GET /city-autocomplete/?q=s&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 610
[29/Apr/2018 17:34:07] "GET /city-autocomplete/?q=Shang&forward=%7B%22FromCountry%22%3Anull%7D HTTP/1.1" 200 418
```

Are these helpful? If not, please elaborate on what I should look for and how I should find them. Thanks!

**Comment by mfxuus** [comment]
@jpic Ahhh I found out - Sorry for wasting your time! I'm not sure where I got this code (probably following some online tutorials a long time ago), but there were these lines in my views.py:

```
# auto complete
class CountryAutocomplete(autocomplete.Select2QuerySetView):
    def get_queryset(self):
        # Don't forget to filter out results depending on the visitor !
        # if not self.request.user.is_authenticated():
        #     return Country.objects.none()

        qs = Country.objects.all()

        if self.q:
            qs = qs.filter(name__istartswith=self.q)

        return qs


class CityAutocomplete(autocomplete.Select2QuerySetView):
    def get_queryset(self):
        # if not self.request.user.is_authenticated():
        #     return Country.objects.none()

        qs = City.objects.all()


        FromCountry = self.forwarded.get('FromCountry', None)
        ToCountry = self.forwarded.get('ToCountry', None)

        if FromCountry:
            qs = qs.filter(country=FromCountry)
        elif ToCountry:
        	qs = qs.filter(country=ToCountry)
        if self.q:
            qs = qs.filter(name__istartswith=self.q)

        return qs
```

So yes, it is actually filtering 'name' with 'startswith'... It has been so long I didn't remember how I implemented this in the beginning. Thanks for looking at it and sorry for bothering you!

**Comment by mpasternak** [comment]
Just came by to tell you exactly that. Glad you've found your solution. Bother us as much as you need, we're the **friendly** hacker club.
